internal/controllers/openshift.go

```diff
@@ -0,0 +1,199 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	operatorv1beta1 "github.com/cryostatio/cryostat-operator/api/v1beta1"
+	"github.com/cryostatio/cryostat-operator/internal/controllers/common"
+	configv1 "github.com/openshift/api/config/v1"
+	consolev1 "github.com/openshift/api/console/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (r *CryostatReconciler) reconcileOpenShift(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	if !r.IsOpenShift {
+		return nil
+	}
+	err := r.reconcileConsoleLink(ctx, cr)
+	if err != nil {
+		return err
+	}
+	return r.addCorsAllowedOriginIfNotPresent(ctx, cr)
+}
+
+func (r *CryostatReconciler) finalizeOpenShift(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	if !r.IsOpenShift {
+		return nil
+	}
+	err := r.deleteConsoleLink(ctx, newConsoleLink(cr))
+	if err != nil {
+		return err
+	}
+	return r.deleteCorsAllowedOrigins(ctx, cr)
+}
+
+func newConsoleLink(cr *operatorv1beta1.Cryostat) *consolev1.ConsoleLink {
+	// Cluster scoped, so use a unique name to avoid conflicts
+	return &consolev1.ConsoleLink{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: common.ClusterUniqueName(cr),
+		},
+	}
+}
+
+func (r *CryostatReconciler) reconcileConsoleLink(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	link := newConsoleLink(cr)
+
+	url := cr.Status.ApplicationURL
+	if len(url) == 0 {
+		// Nothing to link to, so remove the link if it exists
+		err := r.deleteConsoleLink(ctx, link)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, link, func() error {
+		link.Spec.Link = consolev1.Link{
+			Text: "Cryostat",
+			Href: url,
+		}
+		link.Spec.Location = consolev1.NamespaceDashboard
+		link.Spec.NamespaceDashboard = &consolev1.NamespaceDashboardSpec{
+			Namespaces: []string{cr.Namespace},
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	r.Log.Info(fmt.Sprintf("Console Link %s", op), "name", link.Name)
+	return nil
+}
+
+func (r *CryostatReconciler) deleteConsoleLink(ctx context.Context, link *consolev1.ConsoleLink) error {
+	err := r.Client.Delete(ctx, link)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Info("ConsoleLink not found, proceeding with deletion", "name", link.Name)
+			return nil
+		}
+		r.Log.Error(err, "failed to delete ConsoleLink", "name", link.Name)
+		return err
+	}
+	r.Log.Info("deleted ConsoleLink", "name", link.Name)
+	return nil
+}
+
+func (r *CryostatReconciler) addCorsAllowedOriginIfNotPresent(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	reqLogger := r.Log.WithValues("Request.Namespace", cr.Namespace, "Request.Name", cr.Name)
```

Comment:
Can we use `CryostatReconciler.Log` here instead?

Reply:
Same for `deleteCorsAllowedOrigins`?

Reply:
This is using that log, but configuring it so the request name and namespace are added to each log message. This is useful when logging things to do with cluster-scoped resources where it's not obvious which CR caused the log message.

Reply:
Oh yehh from a second look, I understand now. Thanks for explaining :D